CL-900 Remove unused attributes from json schema + table columns + related code

jinjagit opened the pull request:
- [x] Prepared branch for code review

## Description
Removes unused settings from `settings.json.schema` & unused `app_configurations` table columns + related uses in code, following recent creation and use of new `HomePage` model and db table.

## Deployment plan:
@alexander-cit Made some good points around deploying this:

We should take some time to manually test on staging that we are not breaking functionality. Although it's hard to be sure we are aware of all the functionality we might break, we should at least test the related homepage functionality is all still working. 

- [x] First, make sure we have a header banner image set, some images included in the top and bottom info sections of the homepage on staging, and a value set for every setting that relates to the old settings removed in this PR (see list, below).
- [ ] After merging to master, check that the value of every listed field is preserved after the release to staging and then updating them, then it's perfect.
- [ ] Then, check we can remove, replace & update all the elements of the homepage as expected, including the header image and images in the info sections.

Also, we may well find that the FE still tries to make use of (or at least read) some of the fields removed in these 2 PRs. I will spend a little time checking for request errors in local environment, and we should probably look for these on staging and also after eventual release.

### Settings removed:
(`settings = AppConfiguration.first.settings`)
```
settings['events_widget']['enabled']
settings['core']['currently_working_on_text']
settings['core']['display_header_avatars']
settings['core']['custom_onboarding_fallback_message']
settings['core']['header_title']
settings['core']['header_slogan']
AppConfiguration.first.style['signedOutHeaderOverlayColor']
AppConfiguration.first.style['signedOutHeaderOverlayOpacity']
settings['customizable_homepage_banner']['enabled']
settings['customizable_homepage_banner']['layout']
settings['customizable_homepage_banner']['cta_signed_in_customized_button']['text']
settings['customizable_homepage_banner']['cta_signed_in_type']
settings['customizable_homepage_banner']['cta_signed_in_customized_button']['url']
settings['customizable_homepage_banner']['cta_signed_out_customized_button']['text']
settings['customizable_homepage_banner']['cta_signed_out_type']
settings['customizable_homepage_banner']['cta_signed_out_customized_button']['url']
```

### Columns removed:
```
AppConfiguration.first.header_bg.identifier
AppConfiguration.first.homepage_info_multiloc
```

## Priority / urgency
Not urgent - not fixing nor adding anything.

## Links

- [citizenlab-ee PR](https://github.com/CitizenLabDotCo/citizenlab-ee/pull/273)

jinjagit committed:
CL-900 Remove required attributes from json schema to permit deletion of related data

jinjagit committed:
Merge branch 'CL-5_flexible_pages_iteration_1' into CL-900-Remove-old-home-page-data

nTraum commented:
I think we have to merge this into `master`, not `CL-5_flexible_pages_iteration_1`. The epic branch has been merged already.

jinjagit commented:
> I think we have to merge this into `master`, not `CL-5_flexible_pages_iteration_1`. The epic branch has been merged already.

Yes, that's probably true. CL-5 wasn't merged into master when I started this.

jinjagit committed:
[CL-900] Replace required enabled for customizable_homepage_banner in json schema

jinjagit committed:
Merge branch 'master' into CL-900-Remove-old-home-page-data

jinjagit committed:
Merge branch 'master' into CL-900-Remove-old-home-page-data

jinjagit committed:
[CL-900] Remove enabled property from events widget - test

jinjagit committed:
[CL-900] Remove events widget enabled from seed, not allowed

jinjagit committed:
[CL-900] Remove display_header_avatars

jinjagit committed:
[CL-900] Remove various header texts - see commit msg details

currently_working_on_text
custom_onboarding_fallback_message
header_title
header_slogan

jinjagit committed:
Remove customizable_homepage_banner enabled

jinjagit committed:
[CL-900] Remove :validate_customizable_homepage_banner + related file & see what tests fail

jinjagit committed:
[CL-900] Don't include CustomizableHomepageBannerSettings in AppConfig model

jinjagit committed:
[CL-900] Add tests of validations of cta buttons text and url

jinjagit committed:
[CL-900] Remove banner related settings from app_configrations factory

jinjagit committed:
[CL-900] Remove banner related settings from seed file

jinjagit committed:
[CL-900] Remove banner related settings from settings.schema.json.rb

jinjagit committed:
[CL-900] Remove comma from  settings.schema.json.rb

jinjagit committed:
[CL-900] Remove header_bg & homepage_info_multiloc columns

jinjagit committed:
[CL-900] Remove homepage_info_multiloc model validations

jinjagit committed:
[CL-900] Remove homepage_info_multiloc from serializer & side_fx_service

jinjagit committed:
[CL-900] Remove header_bg from model & serializer

jinjagit committed:
[CL-900] Removae useless call to before_update

jinjagit committed:
[CL-900] Remove header_bg and homepage_info multiloc from controller & acceptance spec

jinjagit committed:
[CL-900] Remove header_bg from commercial tenants controller, serializer and acceptance spec

jinjagit committed:
[CL-900] Remove header_bg column from tenants db table

jinjagit committed:
[CL-900] Add header image to citizenlab seed

jinjagit committed:
[CL-900] Remove signedOutHeaderOverlayColor from app_configurations.spec

jinjagit commented on back/app/services/side_fx_app_configuration_service.rb:
```diff
@@ -5,19 +5,9 @@ class SideFxAppConfigurationService
-    return unless app_config.homepage_info_multiloc
-
-    app_config.homepage_info_multiloc = TextImageService.new.swap_data_images app_config, :homepage_info_multiloc
-  end
+  def before_update(app_config, _ = nil) end
```
I left these as no-ops, for now. I need to check where they are called from, if anywhere, as it might be more than one place in the code (and an alias might be used for the class name in the calling line.

jinjagit commented on back/db/migrate/20220831142106_remove_header_bg_from_tenants.rb:
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class RemoveHeaderBgFromTenants < ActiveRecord::Migration[6.1]
+  def change
+    remove_column :tenants, :header_bg, :string
+  end
+end
```
I thought long and hard about removing this, particularly as the `tenant_serializer` talks recreating "... legacy serialized tenants", but in the end was satisfied that:

1. Every production tenant which had a banner image, now has a copy of that image referenced by `HomePage.first.header_bg`.
2. We either seed a header image in our dev environment (if desired), or create one based on a template (if a header image is included in the source `HomePage.first`), or manually set an image to taste when none is provided at tenant creation.

Thus, I can't see any reason that we would need to keep a copy of the homepage header image in the Tenant records. Please correct me if my thinking is wrong!

nTraum replied:
That looks good to me.

jinjagit commented on back/db/schema.rb:
```diff
@@ -1087,7 +1085,6 @@
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
     t.string "logo"
-    t.string "header_bg"
```
See [comment](https://github.com/CitizenLabDotCo/citizenlab/pull/2230/files#r960366946) on deleting this from `tenants`.

jinjagit commented on back/db/seeds/citizenlab.rb:
```diff
@@ -554,7 +547,9 @@ def create_for_current_locales
 # Create settings for Home Page.
-HomePage.create!
+HomePage.create!({
+  header_bg: Rails.root.join('spec/fixtures/header.jpg').open
```
Seeds the OS version with our default banner image, as before we implemented the new `HomePage` model.

jinjagit commented on back/engines/commercial/admin_api/app/serializers/admin_api/tenant_serializer.rb:
```diff
@@ -9,17 +9,13 @@ module AdminApi
   #   AdminApi::TenantSerializer.new(tenant, app_configuration: config)
   class TenantSerializer < ActiveModel::Serializer
-    attributes :id, :name, :host, :settings, :style, :logo, :header_bg, :favicon
+    attributes :id, :name, :host, :settings, :style, :logo, :favicon
```
See [comment](https://github.com/CitizenLabDotCo/citizenlab/pull/2230/files#r960366946) on deleting this from `tenants`.

jinjagit commented on back/engines/commercial/admin_api/app/controllers/admin_api/tenants_controller.rb:
```diff
@@ -94,7 +94,7 @@ def set_tenant
 
     def config_params
-      @config_params ||= params.require(:tenant).permit(:host, :name, :logo, :header_bg, settings: {}, style: {})
+      @config_params ||= params.require(:tenant).permit(:host, :name, :logo, settings: {}, style: {})
```
See [comment](https://github.com/CitizenLabDotCo/citizenlab/pull/2230/files#r960366946) on deleting this from `tenants`.

nTraum commented on back/app/controllers/web_api/v1/avatars_controller.rb:
```diff
@@ -29,7 +29,7 @@ def index
       authorize initiative, :show?
       avatars_service.avatars_for_initiative(initiative, users: users, limit: limit)
     when nil
-      users = User.none unless AppConfiguration.instance.settings.dig('core', 'display_header_avatars')
```
Oh wow, we missed that so far? :sweat_smile: Great find.

jinjagit committed:
[CL-900] Remove uunused params from spec

nTraum commented on back/app/controllers/web_api/v1/app_configurations_controller.rb:
```diff
@@ -10,7 +10,6 @@ def show
   def update
     update_configuration!(app_configuration, config_params)
     side_fx_service = SideFxAppConfigurationService.new
-    side_fx_service.before_update(app_configuration, current_user)
```
I think we should leave this one in here. The method is currently a no op, I'd be surprised if I fill it with content, only to find out it's not being called.

jinjagit replied:
Good call, & replaced.

jinjagit committed:
[CL-900] Replace call to side_fx_service.before_update

nTraum reviewed (approved):
LGTM!

nTraum committed:
Merge branch 'master' into CL-900-Remove-old-home-page-data

# Conflicts:
#	back/app/controllers/web_api/v1/app_configurations_controller.rb
#	back/db/schema.rb
